src/main/java/com/snowflake/kafka/connector/internal/FileNameUtils.java

```diff
@@ -78,11 +89,33 @@ static String brokenRecordFileName(String prefix, long offset, boolean isKey) {
    *
    * @param appName connector name
    * @param table table name
+   * @param topic topic name
    * @param partition partition index
    * @return file prefix
    */
-  static String filePrefix(String appName, String table, int partition) {
-    return appName + "/" + table + "/" + partition + "/";
+  static String filePrefix(String appName, String table, String topic, int partition) {
+    if (partition >= 0x8000) {
+      throw new IllegalArgumentException(
+          String.format("partition id=%d is too large (max=%d)", partition, 0x8000));
+    }
+    BigInteger partitionPart = BigInteger.valueOf(partition);
+    if (!Strings.isNullOrEmpty(topic)) {
+      // if topic is provided as part of the file prefix,
+      // 1. lets calculate stable hash code out of it,
+      // 2. bit shift it by 16 bits left,
+      // 3. add 0x8000 (light up 15th bit as a marker)
+      // 4. add partition id (which should in production use cases never reach a value above 5.000
+      // partitions pers topic).
+      // In theory - we would support 32767 partitions, which is more than
+      byte[] bytes = topic.toUpperCase().getBytes(StandardCharsets.UTF_8);
+      BigInteger hash = BigInteger.valueOf(Crc32C.compute(bytes, 0, bytes.length));
+      partitionPart =
+          hash.abs()
+              .multiply(BigInteger.valueOf(0x10000))
+              .add(BigInteger.valueOf(0x8000))
+              .add(partitionPart);
```

Comment:
Why are we calculating the hash? The change is breaking for some of the customers anyway. Why not write a topic name normally?

Reply:
because of length limitation - AFAIK - the name of the file on stage is limited  - i did come upon a number of around 200 character (can't find it now though). the topic name could be anything up to 250 characters, so adding it as is would create huge file names.